src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -473,9 +574,48 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
     ldpp_dout(dpp, 10) << __LINE__ << " " << __func__ << "(): e->bucket_id=" << e->bucket_id << dendl;
     ldpp_dout(dpp, 10) << __LINE__ << " " << __func__ << "(): e->user=" << e->user << dendl;
     ldpp_dout(dpp, 10) << __LINE__ << " " << __func__ << "(): e->obj_key=" << e->obj_key << dendl;
+
+    auto p = o_entries_map.find(e->key);
+    if (p == o_entries_map.end()) {
+      l.unlock();
+      continue;
+    }
+    if (p->second.second == State::INVALID) {
+      invalid = true;
+    }
     l.unlock();
-    if (!e->key.empty() && (std::difftime(time(NULL), e->creationTime) > interval)) { //if block is dirty and written more than interval seconds ago
-      rgw_user c_rgw_user = e->user;
+  
+    // If the state is invalid, the blocks must be deleted from the cache rather than written to the backend.
+    if (invalid) {
+      ldpp_dout(dpp, 10) << __func__ << "(): State is INVALID; deleting object." << dendl;
+      int ret = -1;
+      if ((ret = cacheDriver->delete_data(dpp, "D#" + e->key, y)) == 0) { // Sam: do we want del or delete_data here?
+	if (!(ret = erase(dpp, e->key, y))) {
+	  ldpp_dout(dpp, 0) << "Failed to delete head policy entry for: " << e->key << ", ret=" << ret << dendl;
+	  continue;
+	}
+      } else {
+	ldpp_dout(dpp, 0) << "Failed to delete head object for: " << e->key << ", ret=" << ret << dendl;
+	continue;
+      }
+
+      ret = delete_data_blocks(dpp, e, y);
+      if (ret == 0) {
+	erase_dirty_object(dpp, e->key, null_yield);
```

Comment:
If you mean to continue after deleting data blocks successfully, shouldn't `continue` be added here?